plugin.py

```diff
@@ -1,180 +1,80 @@
+        # prepare server command
+        cls.server_dir = os.path.join(sublime.cache_path(), __package__)
+        cls.server_jar = os.path.join(cls.server_dir, SERVER_URL.rsplit("/", 1)[1])
+        settings.set("command", ["java", "-jar", cls.server_jar])
 
+        # setup scheme cache directory
+        client_settings = settings.get("settings")
+        client_settings["xml.server.workDir"] = cls.server_dir
+        settings.set("settings", client_settings)
 
-class LspXMLServer(object):
-    binary = None
-    checksum = None
-    ready = False
-    url = None
-    version = None
-    thread = None
+        return settings, file_path
 
     @classmethod
-    def check_binary(cls) -> bool:
-        """Check sha256 hash of downloaded binary.
-
-        Make sure not to run malicious or corrupted code.
-        """
+    def needs_update_or_installation(cls):
         try:
-            with open(cls.binary, "rb") as stream:
+            with open(cls.server_jar, "rb") as stream:
                 checksum = hashlib.sha256(stream.read()).hexdigest()
-                return cls.checksum == checksum
+                return checksum.lower() != SERVER_SHA256.lower()
         except OSError:
             pass
-        return False
-
-    @classmethod
-    def download(cls) -> None:
-        with ActivityIndicator(
-            target=sublime.active_window(),
-            label="Downloading XML language server binary",
-        ):
-            urllib.request.urlretrieve(url=cls.url, filename=cls.binary)
-            cls.ready = cls.check_binary()
-            if not cls.ready:
-                try:
-                    os.remove(cls.binary)
-                except OSError:
-                    pass
-
-        if not cls.ready:
-            sublime.error_message("Error downloading XML server binary!")
-
-        cls.thread = None
+        return True
 
     @classmethod
-    def setup(cls) -> None:
-        if cls.thread or cls.ready:
-            return
-
-        # read server source information
-        filename = "Packages/{}/server.json".format(__package__)
-        server_json = sublime.decode_value(sublime.load_resource(filename))
-
-        cls.version = server_json["version"]
-        cls.url = sublime.expand_variables(server_json["url"], {"version": cls.version})
-        cls.checksum = server_json["sha256"].lower()
-
-        # built local server binary path
-        dest_path = package_cache()
-        cls.binary = os.path.join(dest_path, os.path.basename(cls.url))
-
-        # download server binary on demand
-        cls.ready = cls.check_binary()
-        if not cls.ready:
-            cls.thread = threading.Thread(target=cls.download)
-            cls.thread.start()
-
+    def install_or_update(cls):
         # clear old server binaries
-        for fn in os.listdir(dest_path):
-            fp = os.path.join(dest_path, fn)
+        for file_name in os.listdir(cls.server_dir):
+            if file_name.splitext()[1].lower() != ".jar":
```

Comment:
```
Unable to start lemminx
Traceback (most recent call last):
  File "/Users/me/Library/Application Support/Sublime Text/Packages/LSP/plugin/core/windows.py", line 215, in start_async
    plugin_class.install_or_update()
  File "/Users/me/Library/Application Support/Sublime Text/Packages/LSP-lemminx/plugin.py", line 58, in install_or_update
    if file_name.splitext()[1].lower() != ".jar":
AttributeError: 'str' object has no attribute 'splitext'
```

Reply:
Oups. Unfortunatelly never came to that point to see that failing as ST freezes before.